File: Domain/Models/Category/Category.cs
using System.ComponentModel.DataAnnotations;

namespace Domain.Models.Category;

/// <summary>
/// Represents a category. This is an individual budget item
/// such as I want to save 200 a bucks a month, or have
/// a budget of $600 a month for groceries.
/// </summary>
public class Category
{
    /// <summary>
    /// The <see cref="Category.Id"/>
    /// </summary>
    [Required]
    public Guid Id { get; set; }

    /// <summary>
    /// The <see cref="Category.Title"/>
    /// </summary>
    public required string Title { get; set; }

    /// <summary>
    /// How much to assign to this <see cref="Category"/>
    /// </summary>
    public double Target { get; set; }

    /// <summary>
    /// How much has been spent on this <see cref="Category"/>
    /// </summary>
    public double Outflow { get; set; }

    /// <summary>
    /// How much has been assigned to this <see cref="Category"/>
    /// </summary>
    public double Inflow { get; set; }

    /// <summary>
    /// the <see cref="CategoryGroup.Id"/> of this <see cref="Category"/>
    /// </summary>
    public Guid CategoryGroupId { get; set; }
}

File: client-app/src/models/Category.ts
export class Category {
    id: string;
    title: string;
    target: number;
    outflow: number;
    assigned: number;
    categoryGroupId: string;
    available: number;


    constructor(id: string, name: string, categoryGroupId: string, assigned: number = 0, target: number = 0, outflow: number = 0) {
        this.id = id;
        this.title = name;
        this.assigned = assigned;
        this.target = target;
        this.outflow = outflow;
        this.categoryGroupId = categoryGroupId;
        this.available = this.availableBalance();
    }

    private availableBalance(): number {
        return this.assigned - this.outflow;
    }
}